xfel/merging/application/statistics/deltaccint.py

```diff
@@ -0,0 +1,190 @@
+        diff_sq_[expt_idx] -= (intensity[i] - mean_intensity)**2
+
+      variance[:,hkl_idx] = diff_sq_ / n[:,hkl_idx]
+
+    # N expts (all ranks) x N bins
+    n_bins = resolution_binner.n_bins_used()
+    all_i_sums   = np.zeros((len(all_expt_ids), n_bins), float) # sum of the averaged intensities
+    all_i_n      = np.zeros((len(all_expt_ids), n_bins), int)   # count of the averaged intensities
+    all_var_sums = np.zeros((len(all_expt_ids), n_bins), float) # sums of the variances of the intensities
+
+    # Sum up and reduce the bins
+    for hkl in hkl_set:
+      bin_idx = hkl_resolution_bins[hkl] - 1
+      all_i_sums  [:,bin_idx] += merged[:,hkl_map[hkl]]
+      all_i_n     [:,bin_idx] += 1
+      all_var_sums[:,bin_idx] += variance[:,hkl_map[hkl]]
+
+    total_i_sums   = comm.reduce(all_i_sums,   MPI.SUM, 0)
+    total_i_n      = comm.reduce(all_i_n,      MPI.SUM, 0)
+    total_var_sums = comm.reduce(all_var_sums, MPI.SUM, 0)
+
+    # Broadcast the average intensities
+    if self.mpi_helper.rank == 0:
+      total_i_average_ = total_i_sums / total_i_n
+    else:
+      total_i_average_ = None
+    total_i_average = comm.bcast(total_i_average_, 0)
+
+    # Compute the variance of the average intensities
+    # First, the numerator, the difference between each hkl and the average for that hkl's bin (ommiting each experiment once)
+    diff_sq = np.zeros(merged.shape, float)
+    for hkl in hkl_set:
+      diff_sq[:,hkl_map[hkl]] = (merged[:,hkl_map[hkl]] - total_i_average[:,hkl_resolution_bins[hkl]-1]) ** 2
+
+    # N expts (all ranks) x N bins
+    diff_sq_sum = np.zeros(all_i_sums.shape, float)
+
+    # Complete the numerator for this rank
+    for hkl in hkl_set:
+      diff_sq_sum[:,hkl_resolution_bins[hkl]-1] += diff_sq[:,hkl_map[hkl]]
+
+    total_diff_sq_sum = comm.reduce(diff_sq_sum, MPI.SUM, 0)
+
+    # Report
+    if self.mpi_helper.rank == 0:
+      sigma_sq_y = total_diff_sq_sum / all_i_n
+      sigma_sq_e = total_var_sums / total_i_n
+      deltaccint_st = (sigma_sq_y - (0.5 * sigma_sq_e)) / (sigma_sq_y + sigma_sq_e)
+
+      data = flex.double(np.mean(deltaccint_st, axis=1)) * 100
+      sorted_data = data.select(flex.sort_permutation(data))
+
+      mini, q1, med, q3, maxi = five_number_summary(data)
+      self.logger.main_log("Five number summary")
+      self.logger.main_log("% 8.4f%% min"%mini)
+      self.logger.main_log("% 8.4f%% q1"%q1)
+      self.logger.main_log("% 8.4f%% median"%med)
+      self.logger.main_log("% 8.4f%% q3"%q3)
+      self.logger.main_log("% 8.4f%% max"%maxi)
+      self.logger.main_log("")
+
+      n_worst = min(len(data), 30)
+      worst = sorted_data[-n_worst:]
+      iqr = q3-q1
+
+      self.logger.main_log("Showing ΔCC½ of the worst %d lattices and IQR ratio needed to remove them"%n_worst)
+      self.logger.main_log(" ΔCC½ (%) IQR ratio Lattices removed")
+      for i in range(len(worst)):
+        self.logger.main_log("% 8.4f % 10.1f %d"%(worst[i], (worst[i]-med)/iqr, n_worst-i))
```

Comment:
How about logging the "worst" in reverse order so that the biggest offender appears at the very beginning? I think it would be more intuitive:
```
      for i, value in enumerate(reversed(worst)):
        self.logger.main_log("% 8.4f % 10.1f %d"%(value, (value-med)/iqr, i+1))
```

Reply:
This will produce the following table (props to uncle GPT for reversing 😛 ):
```
Showing ΔCC½ of the worst 30 lattices and IQR ratio needed to remove them
 ΔCC½ (%) IQR ratio Lattices removed
 91.4728      311.1 1
 91.3649      103.5 2
 91.3414       58.2 3
 91.3413       58.1 4
 91.3410       57.6 5
 91.3409       57.3 6
 91.3408       57.1 7
 91.3407       56.9 8
 91.3404       56.3 9
 91.3403       56.2 10
 91.3403       56.1 11
 91.3350       45.9 12
 91.3222       21.3 13
 91.3219       20.7 14
 91.3218       20.5 15
 91.3215       20.1 16
 91.3200       17.1 17
 91.3199       16.9 18
 91.3192       15.5 19
 91.3188       14.8 20
 91.3185       14.2 21
 91.3184       14.1 22
 91.3183       13.8 23
 91.3182       13.7 24
 91.3179       13.1 25
 91.3178       12.9 26
 91.3175       12.4 27
 91.3175       12.3 28
 91.3174       12.1 29
 91.3172       11.7 30
```

Reply:
Also, maybe I misunderstood but isn't this table misleading? We don't get  ΔCC½=91.3172 by removing 30 worst lattices, we get it by removing the 30th lattice only – in which case the last column should say something along the lines of "-th worst lattice removed" instead of "Lattices removed". Otherwise this seems to suggest that removing more and more lattices lowers ΔCC½, which I hope is not the case.

Reply:
Heh, I like this one as it is, but I think it's a matter of taste.